src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSessionMonitor.cs

```diff
@@ -0,0 +1,201 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Microsoft.SqlServer.Management.Sdk.Sfc;
+using Microsoft.SqlServer.Management.XEvent;
+using Microsoft.SqlTools.ServiceLayer.Connection.Contracts;
+using Microsoft.SqlTools.ServiceLayer.Profiler.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.Profiler
+{
+    /// <summary>
+    /// Classs to monitor active profiler sessions
+    /// </summary>
+    public class ProfilerSessionMonitor : IProfilerSessionMonitor
+    {
+        private const int PollingLoopDelay = 1000;
+
+        private object sessionsLock = new object();
+
+        private object listenersLock = new object();
+
+        private Task processorThread = null;
+
+        private Dictionary<string, ProfilerSession> monitoredSessions = new Dictionary<string, ProfilerSession>();
+
+        private List<IProfilerSessionListener> listeners = new List<IProfilerSessionListener>();
+
+        /// <summary>
+        /// Registers a session event listener to receive a callback when events arrive
+        /// </summary>
+        public void AddSessionListener(IProfilerSessionListener listener)
+        {   
+            lock (this.listenersLock) 
+            {
+                this.listeners.Add(listener);
+            }
+        }
+
+        /// <summary>
+        /// Start monitoring the provided sessions
+        /// </summary>
+        public bool StartMonitoringSession(ProfilerSession session)
+        {
+            lock (this.sessionsLock)
+            {
+                // start the monitoring thread 
+                if (this.processorThread == null)
+                {
+                    this.processorThread = Task.Factory.StartNew(ProcessSessions);;
+                }
+
+                if (!this.monitoredSessions.ContainsKey(session.SessionId))
```

Comment:
>monitoredSessions [](start = 26, length = 17)

doesn't need to be concurrentdictionary?

Reply:
no, the dictionary is already inside a lock block.  I looked at concurrentdictionary and prefer to manually handle locking myself for this use case.